skfda/misc/rkhs_product.py

```diff
@@ -0,0 +1,216 @@
+        )
+    else:
+        inv_fd2_matrix = np.asarray(
+            scipy.linalg.lstsq(
+                cov_matrix_1_2.T,
+                data_matrix_2[..., 0].T,
+                cond=cond,
+            )[0],
+        ).T[..., np.newaxis]
+
+    products = (
+        np.transpose(data_matrix_1, axes=(0, 2, 1))
+        @ inv_fd2_matrix
+    )
+    # Remove redundant dimensions
+    return products.reshape(products.shape[0])
+
+
+@rkhs_inner_product.register(FDataBasis, FDataBasis)
+def rkhs_inner_product_fdatabasis(
+    fdata1: FDataBasis,
+    fdata2: FDataBasis,
+    *,
+    cov_function: Callable[[NDArrayFloat, NDArrayFloat], NDArrayFloat],
+) -> NDArrayFloat:
+    """RKHS inner product of two FDataBasis objects.
+
+    In case of using a basis expression, the RKHS inner product can be
+    computed obtaining first a basis representation of the covariance
+    function in the tensor basis. Then, the inverse operator applied to
+    the second term is calculated.
+    In case of using a common basis, that step is done by solving the
+    system given by the inverse of the matrix of coefficients of the
+    covariance function in the tensor basis formed by the two given bases.
+    In case of using different bases, the left inverse of the transposed
+    matrix of coefficients of the covariance function is used instead.
+    Finally, the inner product between each pair of samples is calculated.
+
+    In case of knowing the matrix of coefficients of the covariance function
+    in the tensor basis formed by the two given bases, it can be passed as
+    an argument to avoid the calculation of it.
+
+    Args:
+        fdata1: First FDataBasis object.
+        fdata2: Second FDataBasis object.
+        cov_function: Covariance function as a callable. It is expected to
+            receive two arrays, s and t, and return the corresponding
+            covariance matrix.
+
+    Returns:
+        Matrix of the RKHS inner product between paired samples of
+        fdata1 and fdata2.
+    """
+    _check_compatible_fdata(fdata1, fdata2)
+
+    if isinstance(cov_function, EmpiricalBasis):
+        cov_coeff_matrix = cov_function.coeff_matrix
+    else:
+        # Express the covariance function in the tensor basis
+        # NOTE: The alternative is to convert to FDatagrid the two FDataBasis
+        cov_coeff_matrix = _get_coeff_matrix(
+            cov_function,
+            fdata1.basis,
+            fdata2.basis,
+        )
+
+    if fdata1.basis == fdata2.basis:
+        inv_fd2_coefs = np.linalg.solve(
+            cov_coeff_matrix,
+            fdata2.coefficients.T,
+        )
+    else:
+        inv_fd2_coefs = np.linalg.lstsq(
+            cov_coeff_matrix.T,
+            fdata2.coefficients.T,
+            rcond=None,
+        )[0]
+
+    # Following einsum is equivalent to doing the matrix multiplication
+    # and then taking the diagonal of the result
+    return np.einsum(  # type: ignore[no-any-return]
```

Comment:
Is this faster/more memory efficient than the solution in https://stackoverflow.com/a/14759273/2455333 ? Just out of curiosity.

Reply:
I used `cProfile` to test both methods with 10000 x 10000 matrices and `einsum` is significantly (~14 times) faster. This is mainly because of the `dot` method, since it calculates all products instead of only the diagonal ones.